mods/pvp/grenades/grenades.lua

```diff
@@ -123,78 +128,128 @@ grenades.register_grenade("grenades:frag_sticky", fragdef_sticky)
 local SMOKE_GRENADE_TIME = 30
-grenades.register_grenade("grenades:smoke", {
-	description = "Smoke grenade (Generates smoke around blast site)",
-	image = "grenades_smoke_grenade.png",
-	on_collide = function()
-		return true
-	end,
-	on_explode = function(def, obj, pos, pname)
-		local player = minetest.get_player_by_name(pname)
-		if not player or not pos then return end
 
-		local pteam = ctf_teams.get(pname)
+local register_smoke_grenade = function(name, description, image, damage)
+	grenades.register_grenade("grenades:"..name, {
+		description = description,
+		image = image,
+		on_collide = function()
+			return true
+		end,
+		on_explode = function(def, obj, pos, pname)
+			local player = minetest.get_player_by_name(pname)
+			if not player or not pos then return end
+
+			local pteam = ctf_teams.get(pname)
 
-		if pteam then
-			local fpos = ctf_map.current_map.teams[pteam].flag_pos
+			if pteam then
+				local fpos = ctf_map.current_map.teams[pteam].flag_pos
 
-			if not fpos then return end
+				if not fpos then return end
 
-			if vector.distance(pos, fpos) <= 15 then
-				minetest.chat_send_player(pname, "You can't explode smoke grenades so close to your flag!")
-				player:get_inventory():add_item("main", "grenades:smoke")
-				return
+				if vector.distance(pos, fpos) <= 15 then
+					minetest.chat_send_player(pname, "You can't explode smoke grenades so close to your flag!")
+					player:get_inventory():add_item("main", "grenades:smoke")
+					return
+				end
 			end
-		end
 
-		minetest.sound_play("grenades_glasslike_break", {
-			pos = pos,
-			gain = 1.0,
-			max_hear_distance = 32,
-		})
+			minetest.sound_play("grenades_glasslike_break", {
+				pos = pos,
+				gain = 1.0,
+				max_hear_distance = 32,
+			})
 
-		local hiss = minetest.sound_play("grenades_hiss", {
-			pos = pos,
-			gain = 1.0,
-			loop = true,
-			max_hear_distance = 32,
-		})
-		sounds[hiss] = true
-
-		minetest.after(SMOKE_GRENADE_TIME, function()
-			sounds[hiss] = nil
-			minetest.sound_stop(hiss)
-		end)
-
-		for i = 0, 5, 1 do
-			minetest.add_particlespawner({
-				amount = 40,
-				time = SMOKE_GRENADE_TIME + 3,
-				minpos = vector.subtract(pos, 2),
-				maxpos = vector.add(pos, 2),
-				minvel = {x = 0, y = 2, z = 0},
-				maxvel = {x = 0, y = 3, z = 0},
-				minacc = {x = 1, y = 0.2, z = 1},
-				maxacc = {x = 1, y = 0.2, z = 1},
-				minexptime = 1,
-				maxexptime = 1,
-				minsize = 125,
-				maxsize = 140,
-				collisiondetection = false,
-				collision_removal = false,
-				vertical = false,
-				texture = "grenades_smoke.png^[noalpha",
+			local hiss = minetest.sound_play("grenades_hiss", {
+				pos = pos,
+				gain = 1.0,
+				loop = true,
+				max_hear_distance = 32,
 			})
-		end
-	end,
-	particle = {
-		image = "grenades_smoke.png",
-		life = 1,
-		size = 4,
-		glow = 0,
-		interval = 0.3,
-	}
-})
+			sounds[hiss] = true
+		    if damage then
+				local function damage_fn()
+					local objects = minetest.get_objects_inside_radius(pos, 6)
```

Comment:
```suggestion
		    if damage then
		    	local stop = false
		    	minetest.after(bla, function() stop = true end)
				local function damage_fn()
					if stop then return end
					local objects = minetest.get_objects_inside_radius(pos, 6)
```

You can do something like this to stop the smoke after a certain amount of time. I'd recommend putting the `minetest.after()` next to the initial `damage_fn()` call though

Reply:
(I don't recommend committing this suggestion)

Reply:
Done